File: void-ip-resolve/src/main/kotlin/cn.toutatis.xvoid.resolve/ip/commands/commandLib/dns/IpLib.java
package cn.toutatis.xvoid.resolve.ip.commands.commandLib.dns;

import cn.toutatis.xvoid.resolve.ip.IPResolver;
import cn.toutatis.xvoid.resolve.ip.commands.commandLib.support.CommandHelper;
import cn.toutatis.xvoid.resolve.ip.scan.UrlPoolScanner;
import cn.toutatis.xvoid.toolkit.log.LoggerToolkit;
import com.alibaba.fastjson.JSONArray;
import com.alibaba.fastjson.JSONObject;
import org.slf4j.Logger;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.Charset;
import java.util.Map;

/**
 * @author Toutatis_Gc
 * @date 2022/10/10 16:16
 * IP 相关命令库
 */
public class IpLib extends CommandHelper {

    private final static Logger logger = LoggerToolkit.getLogger(IpLib.class);


    private static String lastRecord = null;

    private static final UrlPoolScanner urlPoolScanner = new UrlPoolScanner(IPResolver.urlPool);

    public static void execute(String target,Object args) {
        JSONArray ipGroup = urlPoolScanner.getIp();
        if (ipGroup!= null && !ipGroup.isEmpty()){
            JSONObject analysisArgs = analysisArgs(args);
            Boolean choosePingIp = (Boolean) analysisArgs.getOrDefault("p", false);
            if (ipGroup.size() == 1){
                Map.Entry<String, Integer> ip = (Map.Entry<String, Integer>) ipGroup.get(0);
                logger.info("当前公网IP为["+ip.getKey()+"]");
                lastRecord = ip.getKey();
                if (choosePingIp){
                    pingIP(lastRecord);
                }
            }else{
                logger.info("当前命中目标"+ipGroup.size()+"个,按优先级排列.");
                for (Object ips : ipGroup) {
                    Map.Entry<String, Integer> ip = (Map.Entry<String, Integer>) ips;
                    logger.info(ip.getKey()+" HITS:"+ip.getValue());
                    String defaultChooseFirstAddress = IPResolver.config.getProperty("Default-Choose-First-Address");
                    if (Boolean.parseBoolean(defaultChooseFirstAddress)){
                        logger.info("当前公网IP为["+ip.getKey()+"]");
                        lastRecord = ip.getKey();
                    }
                }
            }
        }else{
            logger.error("当前未发现公网IP,请检查网络或配置文件.");
        }
    }

    public static String getLastRecord() {
        return lastRecord;
    }


    private static void pingIP(String lastRecord){
        logger.info("WAITING PING URL:"+lastRecord);
        try {
            Process process = Runtime.getRuntime().exec("ping "+lastRecord);
            BufferedReader br = new BufferedReader(new InputStreamReader(process.getInputStream(), Charset.forName("GBK")));
            String line = br.readLine();
            while(line!= null){
                logger.info(line);
                line = br.readLine();
            }
            br.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

}

File: void-ip-resolve/src/main/kotlin/cn.toutatis.xvoid.resolve/ip/commands/commandLib/IpLib.kt
package cn.toutatis.ip.commands.commandLib


import cn.toutatis.ip.commands.commandLib.support.BaseCommand
import cn.toutatis.ip.commands.commandLib.support.CommandHelper
import cn.toutatis.ip.scan.UrlPoolScanner
import cn.toutatis.xvoid.resolve.ip.IPResolver
import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.nio.charset.Charset


object IpScanCmd : BaseCommand,CommandHelper() {

    var lastRecord : String? = null

    private val urlPoolScanner = UrlPoolScanner(IPResolver.urlPool)

    override fun execute(n0: String?, args: Any?) {
        val ipGroup = urlPoolScanner.getIp()
        if (ipGroup != null){
            val analysisArgs = this.analysisArgs(args)
            val choosePingIp = analysisArgs.getOrDefault("p", true) as Boolean
            when(ipGroup.size){
                1 ->{
                    val ip = ipGroup[0] as Map.Entry<*, *>
                    logger.info("当前公网IP为[${ip.key}]")
                    lastRecord = ip.key as String
                    if (choosePingIp){ pingIP(lastRecord!!) }
                }
                else ->{
                    logger.info("当前命中目标${ipGroup.size}个,按优先级排列.")
                    for (ips in ipGroup) {
                        ips as Map.Entry<*, *>
                        logger.info("${ips.key} HITS:${ips.value}")
                    }
                    val defaultChooseFirstAddress = IPResolver.config.getProperty("Default-Choose-First-Address")
                    if(defaultChooseFirstAddress.toBoolean()){
                        lastRecord = ipGroup[0] as String
                        if (choosePingIp){ pingIP(lastRecord!!) }
                    }
                }
            }
        }else{
            logger.error("当前未发现公网IP,请检查网络或配置文件.")
        }

    }

    override fun execute(): Any? {
        TODO("Not yet implemented")
    }

    private fun pingIP(address:String): Unit {
        logger.info("等待 ping 公网IP:${address}")
        val process = Runtime.getRuntime().exec("ping $address")
        val inputStream: InputStream = process.inputStream
        val br = BufferedReader(InputStreamReader(inputStream,Charset.forName("GBK")))
        var line = br.readLine()
        while (line != null) {
            println(line)
            line = br.readLine()
        }
    }

}

object IpRefreshCmd : BaseCommand {

    override fun execute(target: String?, args: Any?) {
        TODO("Not yet implemented")
    }

    override fun execute(): Any? {
        TODO("Not yet implemented")
    }

}